clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -682,8 +683,9 @@ private long partitionReady(Cluster cluster, long nowMs, String topic,
             TopicPartition part = new TopicPartition(topic, entry.getKey());
             // Advance queueSizesIndex so that we properly index available
             // partitions.  Do it here so that it's done for all code paths.
-            Node leader = cluster.leaderFor(part);
-            if (leader != null && queueSizes != null) {
+
+            Metadata.LeaderAndEpoch leaderAndEpoch = metadata.currentLeader(part);
```

Comment:
Since the `leaderAndEpoch` variable isn't reassigned in this method, can we grab a reference to the leader `Node` as we had before? That way we don't have quite so many places that we have the more verbose `leaderAndEpoch.leader.isPresent` and `leaderAndEpoch.leader.get()` everywhere.

Reply:
```suggestion
            Metadata.LeaderAndEpoch leaderAndEpoch = metadata.currentLeader(part);
            Node leader = leaderAndEpoch.orElse(null);
```

Or something like that?